src/xwindows/keyboard.rs

```diff
@@ -0,0 +1,590 @@
+                None,
+            );
+
+            cookie
+                .request_check()
+                .expect("failed to select notify events from xcb xkb");
+        }
+
+        let kbd = Keyboard {
+            context: context,
+            device_id: device_id,
+            keymap: RefCell::new(keymap),
+            state: RefCell::new(state),
+            compose_state: RefCell::new(compose_state),
+        };
+
+        (kbd, first_ev, first_er)
+    }
+
+    pub fn process_key_event(
+        &self,
+        xcb_ev: &xcb::KeyPressEvent,
+    ) -> Option<(KeyCode, KeyModifiers)> {
+        let xcode = xcb_ev.detail() as xkb::Keycode;
+        let xsym = self.state.borrow().key_get_one_sym(xcode);
+        let pressed = (xcb_ev.response_type() & !0x80) == xcb::KEY_PRESS;
+        if pressed {
+            self.compose_state.borrow_mut().feed(xsym);
+        } else {
+            return None;
+        }
+
+        let cstate = self.compose_state.borrow().status().clone();
+        let ksym = match cstate {
+            ComposeStatus::Composing => {
+                // eat
+                return None;
+            }
+            ComposeStatus::Composed => {
+                let res = self.compose_state.borrow().keysym();
+                self.compose_state.borrow_mut().reset();
+                res.unwrap_or(xsym)
+            }
+            ComposeStatus::Nothing => xsym,
+            ComposeStatus::Cancelled => {
+                self.compose_state.borrow_mut().reset();
+                return None;
+            }
+        };
+
+        // could be from_u32_unchecked
+        let ks_char = std::char::from_u32(xkb::keysym_to_utf32(ksym));
+
+        let kc = match ks_char {
+            Some(c) if (c as u32) >= 0x20 && (c as u32) != 0x7f => KeyCode::Char(c),
+            _ => {
+                if let Some(key) = keysym_to_keycode(xsym) {
+                    key
+                } else {
+                    debug!("xkbc:Missing xcb keysym {} definition", xsym);
+                    return None;
+                }
+            }
+        };
+
+        Some((kc, self.get_key_modifiers()))
+    }
+
+    fn mod_is_active(&self, modifier: &str) -> bool {
+        // [TODO] consider state  Depressed & consumed mods
+        self.state
+            .borrow()
+            .mod_name_is_active(modifier, xkb::STATE_MODS_EFFECTIVE)
+    }
+
+    pub fn get_key_modifiers(&self) -> KeyModifiers {
+        let mut res = KeyModifiers::default();
+
+        if self.mod_is_active(xkb::MOD_NAME_SHIFT) {
+            res |= KeyModifiers::SHIFT;
+        }
+        if self.mod_is_active(xkb::MOD_NAME_CTRL) {
+            res |= KeyModifiers::CTRL;
+        }
+        if self.mod_is_active(xkb::MOD_NAME_ALT) {
+            // Mod1
+            res |= KeyModifiers::ALT;
+        }
+        if self.mod_is_active(xkb::MOD_NAME_LOGO) {
+            // Mod4
+            res |= KeyModifiers::SUPER;
+        }
+        if self.mod_is_active("Mod3") {
```

Comment:
Just curious: is this something that we'd want to make configurable in the future? (doesn't need to be part of this PR)

Reply:
yes but keep in mind that it's already configurable in X11 (I rebind caps lock to ```Control``` and ```left_control``` to ```hyper/mod4```).
this is how I see this would work:
  - add a ```HYPER``` variant in ```KeyModifiers```.  
  - keep 1:1 X11 modifiers map to ```KeyModifiers```
  - Allow users to configure based on ```KeyModifiers```